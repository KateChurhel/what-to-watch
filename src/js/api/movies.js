/* istanbul ignore file */
import { sendGetRequest } from './index';

export const getCategoryData = (category, data) => sendGetRequest(`/${category}/popular`, data);

export const getDetailsData = (category, id, data) => sendGetRequest(`/${category}/${id}`, data);

export const getSearchResult = (category, data) => sendGetRequest(`/search/${category}`, data);

export const getDiscoverResult = (data) => sendGetRequest('/discover/movie', data);

export const getGenres = () => sendGetRequest('/genre/movie/list');
