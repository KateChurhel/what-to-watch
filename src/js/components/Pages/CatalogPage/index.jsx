// libraries
import React from 'react';
import PropTypes from 'prop-types';
import { useParams, useLocation } from 'react-router-dom';
// views
import GenresList from '../../Partials/GenresList';
import Pagination from '../../Partials/Pagination';
import MoviesCatalog from '../../Partials/MoviesCatalog';
// constants
import CATEGORIES from '../../../constants/categories';

const CatalogPage = ({
  total_pages: totalPages, total_results: totalResults, results: catalogData = [], isGenresFilterShown = false,
}) => {
  const { category } = useParams();
  const { search } = useLocation();

  let headerText;

  if (search) {
    headerText = `Search Results - ${totalResults} found`;
  } else {
    headerText = `${category === CATEGORIES.movie ? 'Movies' : 'TV Shows'}`;
  }

  return (
    <div className="catalog container ">
      <h2 className="page-title">{headerText}</h2>
      {isGenresFilterShown && (<GenresList />)}
      {catalogData.length > 0 ? (
        <>
          <MoviesCatalog catalogData={catalogData} />
          <Pagination totalPages={totalPages} />
        </>
      ) : (<div className="page-message">No results found</div>)}
    </div>
  );
};

CatalogPage.propTypes = {
  total_pages: PropTypes.number.isRequired,
  total_results: PropTypes.number.isRequired,
  results: PropTypes.arrayOf(PropTypes.object),
  isGenresFilterShown: PropTypes.bool,
};

export default CatalogPage;
