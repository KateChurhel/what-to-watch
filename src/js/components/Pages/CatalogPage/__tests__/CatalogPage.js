import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router-dom';
import { useSelector } from 'react-redux';
import CatalogPage from '../index';
import CATEGORIES from '../../../../constants/categories';
import ROUTES from '../../../../constants/routes';
import { MOCK_MOVIES } from '../../../../mocks/movies';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn(),
}));

const mockData = {
  emptyResult: {
    total_pages: 0,
    total_results: 0,
  },
  withCatalogData: {
    total_pages: 1,
    total_results: 1,
    results: [MOCK_MOVIES[0]],
    isGenresFilterShown: false,
  },
  withFilterAndPagination: {
    total_pages: 2,
    total_results: 1,
    results: MOCK_MOVIES.slice(0, 3),
    isGenresFilterShown: true,
  },
};

describe('Catalog Page: ', () => {
  it('correctly renders with empty result', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/${CATEGORIES.movie}/1`]} initialIndex={1}>
        <Route path={ROUTES.catalog}><CatalogPage {...mockData.emptyResult} /></Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('correctly renders with 1 result', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/${CATEGORIES.movie}/1`]} initialIndex={1}>
        <Route path={ROUTES.catalog}><CatalogPage {...mockData.withCatalogData} /></Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('correctly renders with filter and pagination', async () => {
    useSelector.mockImplementation((selectorFn) => selectorFn({ genres: { activeGenres: [] } }));

    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/${CATEGORIES.tv}/1`]} initialIndex={1}>
        <Route path={ROUTES.catalog}><CatalogPage {...mockData.withFilterAndPagination} /></Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('correctly renders with empty search result', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/search/${CATEGORIES.tv}/1?query=test`]} initialIndex={1}>
        <Route path={ROUTES.search}><CatalogPage {...mockData.emptyResult} /></Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('correctly renders with search result', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/search/${CATEGORIES.movie}/1?query=test`]} initialIndex={1}>
        <Route path={ROUTES.search}><CatalogPage {...mockData.withCatalogData} /></Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
