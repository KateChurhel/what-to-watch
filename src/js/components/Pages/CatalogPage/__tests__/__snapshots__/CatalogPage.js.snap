// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog Page:  correctly renders with 1 result 1`] = `
<DocumentFragment>
  <div
    class="catalog container "
  >
    <h2
      class="page-title"
    >
      Movies
    </h2>
    <div
      class="catalog__movies-list"
    >
      <a
        class="small-movie-card catalog__movies-card small-movie-card__link"
        href="/details/movie/313369"
      >
        <div
          class="small-movie-card__image "
        >
          <img
            alt="La La Land"
            height="450"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w300/uDO8zWDhfWwoFdKS4fzkUJt0Rf0.jpg"
            width="300"
          />
        </div>
        <div
          class="small-movie-card__info"
        >
          <h3
            class="small-movie-card__title"
          >
            La La Land
          </h3>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Catalog Page:  correctly renders with empty result 1`] = `
<DocumentFragment>
  <div
    class="catalog container "
  >
    <h2
      class="page-title"
    >
      Movies
    </h2>
    <div
      class="page-message"
    >
      No results found
    </div>
  </div>
</DocumentFragment>
`;

exports[`Catalog Page:  correctly renders with empty search result 1`] = `
<DocumentFragment>
  <div
    class="catalog container "
  >
    <h2
      class="page-title"
    >
      Search Results - 0 found
    </h2>
    <div
      class="page-message"
    >
      No results found
    </div>
  </div>
</DocumentFragment>
`;

exports[`Catalog Page:  correctly renders with filter and pagination 1`] = `
<DocumentFragment>
  <div
    class="catalog container "
  >
    <h2
      class="page-title"
    >
      TV Shows
    </h2>
    <div
      class="catalog__movies-list"
    >
      <a
        class="small-movie-card catalog__movies-card small-movie-card__link"
        href="/details/tv/313369"
      >
        <div
          class="small-movie-card__image "
        >
          <img
            alt="La La Land"
            height="450"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w300/uDO8zWDhfWwoFdKS4fzkUJt0Rf0.jpg"
            width="300"
          />
        </div>
        <div
          class="small-movie-card__info"
        >
          <h3
            class="small-movie-card__title"
          >
            La La Land
          </h3>
        </div>
      </a>
      <a
        class="small-movie-card catalog__movies-card small-movie-card__link"
        href="/details/tv/516486"
      >
        <div
          class="small-movie-card__image "
        >
          <img
            alt="Greyhound"
            height="450"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w300/kjMbDciooTbJPofVXgAoFjfX8Of.jpg"
            width="300"
          />
        </div>
        <div
          class="small-movie-card__info"
        >
          <h3
            class="small-movie-card__title"
          >
            Greyhound
          </h3>
        </div>
      </a>
      <a
        class="small-movie-card catalog__movies-card small-movie-card__link"
        href="/details/tv/385103"
      >
        <div
          class="small-movie-card__image "
        >
          <img
            alt="Scoob!"
            height="450"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w300/jHo2M1OiH9Re33jYtUQdfzPeUkx.jpg"
            width="300"
          />
        </div>
        <div
          class="small-movie-card__info"
        >
          <h3
            class="small-movie-card__title"
          >
            Scoob!
          </h3>
        </div>
      </a>
    </div>
    <ul
      class="pagination"
    >
      <li>
        <span
          class="pagination__page-number pagination__active-page"
        >
          1
        </span>
      </li>
      <li>
        <a
          class="pagination__page-number"
          href="/tv/2"
        >
          2
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Catalog Page:  correctly renders with search result 1`] = `
<DocumentFragment>
  <div
    class="catalog container "
  >
    <h2
      class="page-title"
    >
      Search Results - 1 found
    </h2>
    <div
      class="catalog__movies-list"
    >
      <a
        class="small-movie-card catalog__movies-card small-movie-card__link"
        href="/details/movie/313369"
      >
        <div
          class="small-movie-card__image "
        >
          <img
            alt="La La Land"
            height="450"
            loading="lazy"
            src="https://image.tmdb.org/t/p/w300/uDO8zWDhfWwoFdKS4fzkUJt0Rf0.jpg"
            width="300"
          />
        </div>
        <div
          class="small-movie-card__info"
        >
          <h3
            class="small-movie-card__title"
          >
            La La Land
          </h3>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
